android_jni/avifandroidjni/src/main/jni/libavif_jni.cc

```diff
@@ -124,66 +205,109 @@ FUNC(jboolean, decode, jobject encoded, int length, jobject bitmap,
+  const jfieldID alpha_present_id =
+      env->GetFieldID(avif_decoder_class, "alphaPresent", "Z");
+  const jfieldID frame_count_id =
+      env->GetFieldID(avif_decoder_class, "frameCount", "I");
+  const jfieldID repetition_count_id =
+      env->GetFieldID(avif_decoder_class, "repetitionCount", "I");
+  const jfieldID frame_durations_id =
+      env->GetFieldID(avif_decoder_class, "frameDurations", "[D");
+  env->SetIntField(thiz, width_id, decoder->decoder->image->width);
+  env->SetIntField(thiz, height_id, decoder->decoder->image->height);
+  env->SetIntField(thiz, depth_id, decoder->decoder->image->depth);
+  env->SetBooleanField(thiz, alpha_present_id, decoder->decoder->alphaPresent);
+  env->SetIntField(thiz, repetition_count_id,
+                   decoder->decoder->repetitionCount);
+  const int frameCount = decoder->decoder->imageCount;
+  env->SetIntField(thiz, frame_count_id, frameCount);
+  // This native array is needed because setting one element at a time to a Java
+  // array from the JNI layer is inefficient.
+  std::unique_ptr<double[]> native_durations(
+      new (std::nothrow) double[frameCount]);
+  if (native_durations == nullptr) {
+    return 0;
   }
-  avifRGBImage rgb_image;
-  avifRGBImageSetDefaults(&rgb_image, decoder.decoder->image);
-  if (bitmap_info.format == ANDROID_BITMAP_FORMAT_RGBA_F16) {
-    rgb_image.depth = 16;
-    rgb_image.isFloat = AVIF_TRUE;
-  } else if (bitmap_info.format == ANDROID_BITMAP_FORMAT_RGB_565) {
-    rgb_image.format = AVIF_RGB_FORMAT_RGB_565;
-    rgb_image.depth = 8;
-  } else {
-    rgb_image.depth = 8;
+  for (int i = 0; i < frameCount; ++i) {
+    avifImageTiming timing;
+    if (avifDecoderNthImageTiming(decoder->decoder, i, &timing) !=
+        AVIF_RESULT_OK) {
+      return 0;
+    }
+    native_durations[i] = timing.duration;
   }
-  rgb_image.pixels = static_cast<uint8_t*>(bitmap_pixels);
-  rgb_image.rowBytes = bitmap_info.stride;
-  // Android always sees the Bitmaps as premultiplied with alpha when it renders
-  // them:
-  // https://developer.android.com/reference/android/graphics/Bitmap#setPremultiplied(boolean)
-  rgb_image.alphaPremultiplied = AVIF_TRUE;
-  res = avifImageYUVToRGB(decoder.decoder->image, &rgb_image);
-  AndroidBitmap_unlockPixels(env, bitmap);
-  if (res != AVIF_RESULT_OK) {
-    LOGE("Failed to convert YUV Pixels to RGB. Status: %d", res);
-    return false;
+  jdoubleArray durations = env->NewDoubleArray(frameCount);
+  if (durations == nullptr) {
+    return 0;
   }
-  return true;
+  env->SetDoubleArrayRegion(durations, /*start=*/0, frameCount,
+                            native_durations.get());
+  env->SetObjectField(thiz, frame_durations_id, durations);
+  return reinterpret_cast<jlong>(decoder.release());
+}
+
+FUNC(jint, nextFrame, jlong jdecoder, jobject bitmap) {
+  AvifDecoderWrapper* const decoder =
+      reinterpret_cast<AvifDecoderWrapper*>(jdecoder);
+  return DecodeNextImage(env, decoder, bitmap);
```

Comment:
Should we check `decoder != 0` before calling `DecodeNextImage()`?

Reply:
Because of the way the java class is structured (the factory method returns null if the decoder creation fails), it is impossible to get here if decoder == 0.

Reply:
Right, I missed the fact that the ctor is private.